crates/bevy_pbr/src/light_probe/environment_filter.wgsl

```diff
@@ -0,0 +1,427 @@
+            face = 1u; // -X
+            uv = vec2f(dir.z, -dir.y) / abs_dir.x;
+        }
+    } else if (abs_dir.y >= abs_dir.x && abs_dir.y >= abs_dir.z) {
+        // Y axis is dominant
+        if (dir.y > 0.0) {
+            face = 2u; // +Y
+            uv = vec2f(dir.x, dir.z) / dir.y;
+        } else {
+            face = 3u; // -Y
+            uv = vec2f(dir.x, -dir.z) / abs_dir.y;
+        }
+    } else {
+        // Z axis is dominant
+        if (dir.z > 0.0) {
+            face = 4u; // +Z
+            uv = vec2f(dir.x, -dir.y) / dir.z;
+        } else {
+            face = 5u; // -Z
+            uv = vec2f(-dir.x, -dir.y) / abs_dir.z;
+        }
+    }
+    
+    // Convert from [-1,1] to [0,1]
+    return CubeUV(uv * 0.5 + 0.5, face);
+}
+
+// Sample an environment map with a specific LOD
+fn sample_environment(dir: vec3f, level: f32) -> vec4f {
+    let cube_uv = dir_to_cube_uv(dir);
+    return textureSampleLevel(input_texture, input_sampler, cube_uv.uv, cube_uv.face, level);
+}
+
+// Calculate tangent space for the given normal
+fn calculate_tangent_frame(normal: vec3f) -> mat3x3f {
+    // Use a robust method to pick a tangent
+    var up = vec3f(1.0, 0.0, 0.0);
+    if abs(normal.z) < 0.999 {
+        up = vec3f(0.0, 0.0, 1.0);
+    }
+    let tangent = normalize(cross(up, normal));
+    let bitangent = cross(normal, tangent);
+    return mat3x3f(tangent, bitangent, normal);
+}
+
+// Hammersley sequence for quasi-random points
+fn hammersley_2d(i: u32, n: u32) -> vec2f {
+    // Van der Corput sequence
+    var bits = i;
+    bits = (bits << 16u) | (bits >> 16u);
+    bits = ((bits & 0x55555555u) << 1u) | ((bits & 0xAAAAAAAAu) >> 1u);
+    bits = ((bits & 0x33333333u) << 2u) | ((bits & 0xCCCCCCCCu) >> 2u);
+    bits = ((bits & 0x0F0F0F0Fu) << 4u) | ((bits & 0xF0F0F0F0u) >> 4u);
+    bits = ((bits & 0x00FF00FFu) << 8u) | ((bits & 0xFF00FF00u) >> 8u);
+    let vdc = f32(bits) * 2.3283064365386963e-10; // 1 / 0x100000000
+    return vec2f(f32(i) / f32(n), vdc);
+}
+
+// Blue noise randomization
+fn blue_noise_offset(pixel_coords: vec2u) -> vec2f {
+    // Get a stable random offset for this pixel
+    let noise_size = vec2u(u32(constants.blue_noise_size.x), u32(constants.blue_noise_size.y));
+    let noise_coords = pixel_coords % noise_size;
+    return textureSampleLevel(blue_noise_texture, input_sampler, vec2f(noise_coords) / constants.blue_noise_size, 0.0).rg;
+}
+
+// GGX/Trowbridge-Reitz normal distribution function (D term)
+fn D_GGX(roughness: f32, NdotH: f32) -> f32 {
+    let oneMinusNdotHSquared = 1.0 - NdotH * NdotH;
+    let a = NdotH * roughness;
+    let k = roughness / (oneMinusNdotHSquared + a * a);
+    let d = k * k * (1.0 / PI);
+    return d;
+}
+
+// Importance sample GGX normal distribution function for a given roughness
+fn importance_sample_ggx(xi: vec2f, roughness: f32, normal: vec3f) -> vec3f {
+    // Use roughness^2 to ensure correct specular highlights
+    let a = roughness * roughness;
```

Comment:
I forget which is which, but we need to be careful of using roughness vs perceptual_roughness in this code.

Reply:
Fixed